kura/org.eclipse.kura.db.sqlite.provider/src/main/java/org/eclipse/kura/internal/db/sqlite/provider/SqliteMessageStoreImpl.java

```diff
@@ -0,0 +1,452 @@
+        return withConnection(c -> {
+            try (final PreparedStatement stmt = c.prepareStatement(sql)) {
+                if (params != null) {
+                    for (int i = 0; i < params.length; i++) {
+                        stmt.setInt(2 + i, params[i]); // timeInterval
+                    }
+                }
+
+                try (final ResultSet rs = stmt.executeQuery()) {
+                    return buildStoredMessagesNoPayload(rs);
+                }
+            }
+        }, "Cannot list messages");
+    }
+
+    private synchronized void execute(String sql, Object... params) throws KuraStoreException {
+        withConnection(c -> {
+            try (final PreparedStatement stmt = c.prepareStatement(sql)) {
+                for (int i = 0; i < params.length; i++) {
+                    stmt.setObject(1 + i, params[i]);
+                }
+                stmt.execute();
+                c.commit();
+                return null;
+            }
+        }, "Cannot execute query");
+    }
+
+    private void createIndex(String indexname, String table, String order) throws KuraStoreException {
+        execute("CREATE INDEX IF NOT EXISTS " + indexname + " ON " + table + " " + order + ";");
+        logger.debug("Index {} created, order is {}", indexname, order);
+    }
+
+    // ------------------------------------------------------------------
+    //
+    // Private Methods: Connection Management
+    //
+    // ------------------------------------------------------------------
+
+    private List<StoredMessage> buildStoredMessagesNoPayload(ResultSet rs) throws SQLException {
+        List<StoredMessage> messages = new ArrayList<>();
+        while (rs.next()) {
+            messages.add(buildStoredMessage(rs, false));
+        }
+        return messages;
+    }
+
+    private StoredMessage buildStoredMessage(ResultSet rs, final boolean includePayload) throws SQLException {
+        StoredMessage.Builder builder = new StoredMessage.Builder(rs.getInt("id"))
+                .withTopic(rs.getString(TOPIC_ELEMENT)).withQos(rs.getInt("qos")).withRetain(rs.getBoolean("retain"))
+                .withCreatedOn(rs.getTimestamp("createdOn", this.utcCalendar))
+                .withPublishedOn(rs.getTimestamp("publishedOn", this.utcCalendar))
+                .withConfirmedOn(rs.getTimestamp("confirmedOn", this.utcCalendar)).withPriority(rs.getInt("priority"))
+                .withDroppedOn(rs.getTimestamp("droppedOn"));
+
+        if (includePayload) {
+            builder = builder.withPayload(rs.getBytes("payload"));
+        }
+
+        final String sessionId = rs.getString("sessionId");
+
+        if (sessionId != null) {
+            builder = builder
+                    .withDataTransportToken(new DataTransportToken(rs.getInt("publishedMessageId"), sessionId));
+        }
+
+        return builder.build();
+    }
+
+    private <T> T withConnection(final ConnectionCallable<T> callable, final String exceptionMessage)
```

Comment:
Does it make sense to have a "`withPreparedStatement`" instead of `withConnection` so to factor in some more common code?

Reply:
Something like:
```java
private interface PreparedStatementCallable<T> {
    public T call(PreparedStatement preparedStatement) throws SQLException;
}

private <T> T withPreparedStatement(String sql, String exceptionMessage,
            final PreparedStatementCallable<T> callable) throws KuraStoreException {
    try (final Connection connection = this.dbService.getConnection()) {
        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            return callable.call(preparedStatement);
        }
    } catch (final Exception e) {
        throw new KuraStoreException(e, exceptionMessage);
    }
}
```

Reply:
I have added this to a utility class